core/pta/attestation.c

```diff
@@ -0,0 +1,589 @@
+#include <string.h>
+#include <string_ext.h>
+#include <malloc.h>
+#include <tee_api_types.h>
+#include <crypto/crypto.h>
+#include <crypto/crypto_impl.h>
+#include "attestation.h"
+
+#define RSA_SIG_SIZE_BYTE 256
+
+/*
+ * Attestation data is static to keep until the device boots again.
+ * Afterwards the attestation data has to be reloaded and everthing
+ * has to be attestated newly.
+ */
+static struct attestation_data {
+	uint8_t *iv;
+	uint32_t iv_size;
+	uint8_t *pub;
+	uint32_t pub_size;
+	uint8_t *enc_priv;
+	uint32_t enc_priv_size;
+	uint8_t *sig;
+	uint32_t sig_size;
+} att_data;
+
+static TEE_Result free_att_data(TEE_Result reason)
+{
+	att_data.iv_size = 0;
+	att_data.pub_size = 0;
+	att_data.enc_priv_size = 0;
+	att_data.sig_size = 0;
+
+	if (att_data.iv)
+		free(att_data.iv);
+
+	if (att_data.pub)
+		free(att_data.pub);
+
+	if (att_data.enc_priv)
+		free(att_data.enc_priv);
+
+	if (att_data.sig)
+		free(att_data.sig);
+
+	return reason;
+}
+
+static TEE_Result priv_asn1_decode(uint8_t *priv, const size_t priv_len,
+				   uint8_t **d, size_t *d_size)
+{
+	uint32_t size = 0;
+	uint32_t offset = 7; // Beginn at the start of the modulus
+
+	if (priv_len < offset + 3)
+		return TEE_ERROR_SECURITY;
+
+	// Check the header
+	if (priv[0] != 0x30 ||
+	    priv[1] != 0x82 ||
+	    priv[offset] != 0x02 ||
+	    priv[offset + 1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	// Get the size of the first integer aka modulus
+	size = (priv[offset + 2] << 8) | priv[offset + 3];
+	offset += size + 4;
+
+	// Get the size of the second integer aka exponent
+	if (priv[offset] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	size = priv[offset + 1];
+	offset += size + 2;
+
+	if (priv_len < offset + 3)
+		return TEE_ERROR_SECURITY;
+
+	// Decoding d
+	if (priv[offset] != 0x02 || priv[offset + 1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (priv[offset + 2] << 8) | priv[offset + 3];
+
+	if (priv_len < offset + 4 + size)
+		return TEE_ERROR_SECURITY;
+
+	*d = &priv[offset + 4];
+	*d_size = size;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pub_asn1_decode(uint8_t **n, uint8_t **e,
+				  size_t *n_size, size_t *e_size)
+{
+	uint32_t size = 0;
+
+	// Decoding header
+	if (att_data.pub[0] != 0x30 || att_data.pub[1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[2] << 8) | att_data.pub[3];
+	if ((size + 4) != att_data.pub_size)
+		return TEE_ERROR_SECURITY;
+
+	// Decoding first integer aka modulus
+	if (att_data.pub[4] != 0x02 || att_data.pub[5] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[6] << 8) | att_data.pub[7];
+	*n = &att_data.pub[8];
+	*n_size = size;
+
+	// Decoding second integer aka exponent
+	if (att_data.pub[size + 8] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	*e_size = att_data.pub[size + 9];
+	*e = &att_data.pub[size + 10];
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak get_sys_measurement(uint8_t *ptr)
+{
+	// TODO: Dummy function
+	// get_sys_measurement -> do something hardware specific
```

Comment:
Measurement as in measured boot in case you have a TPM?

Reply:
Yes, exactly. But we do not have a TPM in mind. We would prefer to provision the Chips (Processor or FPGA) and then boot them with secure boot/HAB into a secure state.